examples/asset/remote_asset_loader.rs

```diff
@@ -0,0 +1,167 @@
+//! wasm or Android, since those platforms doesn't have [`AssetWriter`].
+
+use bevy::{
+    asset::io::{
+        AssetReader, AssetReaderError, AssetSource, AssetSourceId, AssetWriter, PathStream, Reader,
+    },
+    prelude::*,
+    utils::BoxedFuture,
+};
+use bevy_internal::asset::io::AssetWriterError;
+use futures_lite::AsyncRead;
+use std::{path::Path, sync::Arc};
+
+/// A remote asset loader implementation that will try to read asset local and, if not found,
+/// download it remotely from a CDN.
+struct RemoteAssetLoader {
+    cdn_prefix: String,
+    reader: Box<dyn AssetReader>,
+    writer: Box<dyn AssetWriter>,
+}
+
+impl RemoteAssetLoader {
+    fn new() -> Self {
+        // Create the root directory if it doesn't exists on local disk.
+        let create_root_dir = true;
+        let writer = AssetSource::get_default_writer("assets".to_string())(create_root_dir)
+            .expect("Current platform doesn't support asset writing.");
+        let reader = AssetSource::get_default_reader("assets".to_string())();
+
+        // This is the CDN which will be downloaded the font when missing localy.
+        let cdn_prefix =
+            "https://github.com/google/fonts/raw/main/ofl/protestrevolution/".to_string();
+
+        RemoteAssetLoader {
+            reader,
+            writer,
+            cdn_prefix,
+        }
+    }
+
+    /// Remote download the asset from CDN.
+    async fn read_remote<'a>(
+        &'a self,
+        path: &'a Path,
+    ) -> Result<Box<dyn AsyncRead + Send + Sync + Unpin + 'a>, AssetReaderError> {
+        // Only the asset name is needed, not the folder.
+        let asset_name = path.file_name().unwrap().to_str().unwrap();
+        let mut url = self.cdn_prefix.clone();
+        url.push_str(asset_name);
+
+        // A simple GET request is used, but you could set custom headers, auth and so on.
+        let request = ehttp::Request::get(url);
+
+        let body = match ehttp::fetch_async(request).await {
+            Ok(response) => {
+                // Since this is an example, only check for 200 status, but in a real world use
+                // it would be wise to check for others 2xx or 3xx status.
+                if response.status != 200 {
+                    return Err(AssetReaderError::HttpError(response.status));
+                }
+
+                response.bytes
+            }
+            Err(error) => {
+                warn!("Failed to read remote asset: {error}");
+                return Err(AssetReaderError::HttpError(500));
+            }
+        };
+
+        // Try to save the downloaded asset on disk.
+        if let Err(AssetWriterError::Io(error)) = self.writer.write_bytes(path, &body).await {
+            return Err(AssetReaderError::Io(Arc::new(error)));
+        }
+
+        // Try to read again the asset, since its saved on disk now.
+        self.reader.read(path).await
+    }
+}
+
+impl AssetReader for RemoteAssetLoader {
```

Comment:
Since this is achieved using `AssetReader` instead of `AssetLoader`, perhaps the `RemoteAssetLoader` type should be named `RemoteAssetReader` for clarity? Likewise for `RemoteAssetLoaderPlugin`

Reply:
This was my first thought, but since It also uses `AssetWriter` I think this could be misleading. Its an `AssetReader` which also contains an `AssetWriter`.

Reply:
That is a reasonable concern I hadn't considered. Perhaps `RemoteAssetProxy` might be a way to avoid any name collisions? Proxy seems like a good description of what this does.

Reply:
I don't like `Proxy` since it doesn't nothing like a Proxy usually does, even so since we are talking about HTTP requests. Let's see if someone else has a different opinion on that subject. I still think `RemoteAssetLoader` is a good name.